plugins/hls-explicit-record-fields-plugin/src/Ide/Plugin/ExplicitFields.hs

```diff
@@ -0,0 +1,266 @@
+    mkCodeAction nfp exts pragma rec = InR CodeAction
+      { _title = mkCodeActionTitle exts
+      , _kind = Just CodeActionRefactorRewrite
+      , _diagnostics = Nothing
+      , _isPreferred = Nothing
+      , _disabled = Nothing
+      , _edit = Just $ mkWorkspaceEdit nfp edits
+      , _command = Nothing
+      , _xdata = Nothing
+      }
+      where
+        edits = catMaybes [ mkTextEdit rec , pragmaEdit ]
+
+        mkTextEdit :: RenderedRecordInfo -> Maybe TextEdit
+        mkTextEdit (RenderedRecordInfo ss r) = TextEdit <$> srcSpanToRange ss <*> pure r
+
+        -- NOTE(ozkutuk): `RecordPuns` extension is renamed to `NamedFieldPuns`
+        -- in GHC 9.4, but we still want to insert `NamedFieldPuns` in pre-9.4
+        -- GHC as well, hence the replacement.
+        -- https://gitlab.haskell.org/ghc/ghc/-/merge_requests/6156
+        pragmaEdit :: Maybe TextEdit
+        pragmaEdit = if NamedFieldPuns `elem` exts
+                       then Nothing
+                       else Just $ patchExtName $ insertNewPragma pragma NamedFieldPuns
+          where
+            patchExtName = L.newText %~ T.replace "Record" "NamedField"
+
+    mkWorkspaceEdit :: NormalizedFilePath -> [TextEdit] -> WorkspaceEdit
+    mkWorkspaceEdit nfp edits = WorkspaceEdit changes Nothing Nothing
+      where
+        changes = Just $ HashMap.singleton (fromNormalizedUri (normalizedFilePathToUri nfp)) (List edits)
+
+    mkCodeActionTitle :: [Extension] -> Text
+    mkCodeActionTitle exts =
+      if NamedFieldPuns `elem` exts
+        then title
+        else title <> " (needs extension: NamedFieldPuns)"
+        where
+          title = "Expand record wildcard"
+
+collectRecordsRule :: Recorder (WithPriority Log) -> Rules ()
+collectRecordsRule recorder = define (cmapWithPrio LogShake recorder) $ \CollectRecords nfp -> do
+  tmr <- use TypeCheck nfp
+  let exts = getEnabledExtensions <$> tmr
+      recs = getRecords <$> tmr
+      renderedRecs = mapMaybe renderRecordInfo <$> recs
+  pure ([], CRR <$> renderedRecs <*> exts)
+
+getEnabledExtensions :: TcModuleResult -> [GhcExtension]
+getEnabledExtensions = map GhcExtension . toList . extensionFlags . ms_hspp_opts . pm_mod_summary . tmrParsed
+
+getRecords :: TcModuleResult -> [RecordInfo]
+getRecords (tmrRenamed -> (hs_valds -> valBinds,_,_,_)) =
+  collectRecords valBinds
+
+data CollectRecords = CollectRecords
+                    deriving (Eq, Show, Generic)
+
+instance Hashable CollectRecords
+instance NFData CollectRecords
+
+data CollectRecordsResult = CRR
+  { recordInfos       :: ![RenderedRecordInfo]
+  , enabledExtensions :: ![GhcExtension]
+  }
+  deriving (Generic)
+
+instance NFData CollectRecordsResult
+
+instance Show CollectRecordsResult where
+  show _ = "<CollectRecordsResult>"
+
+type instance RuleResult CollectRecords = CollectRecordsResult
+
+newtype GhcExtension = GhcExtension { unExt :: Extension }
```

Comment:
... why? Maybe I'll find out by inferring backwards, but you know already, so write it down in the haddock!

Reply:
I newtype'd it so that I can provide an `NFData` instance (which is required for types returned from rules) without creating an orphan instance. Do you think there is a better way to do this? If this is sensible, I can put this rationale in a comment.

Reply:
No, that's fine, it just needs a comment explaining that.